ksqldb-execution/src/main/java/io/confluent/ksql/execution/util/ExpressionTypeManager.java

```diff
@@ -94,14 +101,25 @@ public ExpressionTypeManager(
   }
 
   public SqlType getExpressionSqlType(final Expression expression) {
+    return getExpressionSqlType(expression, Collections.emptyMap(), Collections.emptyList());
+  }
+
+  public SqlType getExpressionSqlType(
+      final Expression expression, final Map<String, SqlType> inputMapping, final List<SqlType> inputTypes
+  ) {
     final ExpressionTypeContext expressionTypeContext = new ExpressionTypeContext();
+    expressionTypeContext.setLambdaTypes(inputMapping);
```

Comment:
For my own understanding: when we call setLambdaTypes here, is the mapping already populated or not? If yes, then we would not need to `mapInputTypes` anymore; if not, why?

Reply:
The mapping is already set here, but it's set in either `CodeGenRunner` or `SqlToJavaVisitor`. this is a way to pass the mapping we've already done in a `TypeContext` into a new `ExpressionTypeContext` that's used to resolve ultimately resolve the types within `ExpressionTypeManager`. `ExpressionTypeManager` is used by both `SqlToJavaVisitor` and `CodeGenRunner` to do type resolution.

I think this ties into what @stevenpyzhang commented earlier, that it may make sense to combine `ExpressionTypeContext` and `TypeContext`, as it would make passing this information infinitely easier. I think I'll go ahead and do that unless @agavra has any objections.

Currently, the difference between the contexts is one, `ExpressionTypeContext`, is just used in `ExpressionTypeManager`. This existed before lambdas for passing around sql return types. `TypeContext`, as the code stands now, is used by `SqlToJavaVisitor` and `CodeGenRunner` to store information relevant to lambdas and pass it to the relevant methods. It doesn't contain the basic `sqlType` storage that `ExpressionTypeManager` does. `ExpressionTypeManager` doesn't currently use `TypeContext` at all. 

For lambdas, the generic -> sqlType population for the variables (x in x => ucase(x)) happens when the context (either `TypeContext` or `ExpressionTypeContext`) is passed information in the `visitLambdaExpression` function in `ExpressionTypeManager`, `CodeGenRunner`, and `SqlToJavaVisitor`. This happens after `visitor.process()` for each of the individual flows through one of the three relevant classes.

Let me know if this was more confusing than enlightening, I'm happy to have another go at explaining the relationship between all these pieces

Reply:
Haven't looked at the code yet, but from your description combining the two makes sense to me!

Reply:
Thanks @lct45 , I think my confusion was that:

If in line 111 here, which is before we start visiting the node, the genericName -> sqlType `inputMapping` we are set via `expressionTypeContex.setLambdaTypes()` in is already *fully* populated, then inside `visitLambdaExpression` (line 200) which is part of the visiting traversal, we should not need to call `mapInputTypes` which is trying to build the mapping again.

Now I think I get your point, which is that the `inputMapping` passed in via `getExpressionSqlType` is not fully populated, i.e. it only contains the mapping entries for variables outside the lambda function, but not contain entries for variables that are only declared inside the lambda function. Is that right?

Reply:
In that case, if we consolidate these two into a single `TypeContext`, would a single `TypeContext` object be used in multiple lambda expressions? I'm asking this because if that's the case, then its `inputTypes` and `lambdaTypeMapping` would be a global structure containing mappings for multiple expressions, and we cannot enforce `inputTypes.size() == argumentList.size()` in `mapInputTypes`.

Reply:
@stevenpyzhang may be able to weigh in more here, but I think all the lambdas inside of a UDF should have the same number of inputs so I don't think there would be a concern if the TypeContext is shared between them. The example I'm thinking of is from the KLIP, `transform_map(map, (k, v) => new_k, (k, v) => new_v)`, so having a global map of `k,v` should be fine. I did just run through it though, and the context is shared. This is necessary for the input though, since in a lambda it'll be one input mapped to the lambda variables (i.e. we only see the map once even if we have 2 lamda expressions). This does raise the issue of variable names. Should users be able to do a lambda like `transform_map(map, (x,y) => x + y, (k,v) => k - v)`? Because I think you're right @guozhangwang , right now I don't  it would accept that

Reply:
I think `transform_map(map, (x,y) => x + y, (k,v) => k - v)` should be supported. We could just make a copy of the TypeContext and pass a fresh instance to each argument that we visit so that we can have different mappings for each Lambda node.